docs/contributing/writing-instrumentation.md

```diff
@@ -46,85 +46,314 @@ Within the subfolder, create three folders `library` (skip if library instrument
+initialized. In a library instrumentation test, there will be code calling into the instrumentation
+API, while in a javaagent instrumentation test it will generally just use the underlying library's
+API as is and just rely on the javaagent to apply all the necessary bytecode changes automatically.
+
+You can use either JUnit 5 (recommended) or Spock to test the instrumentation. In any case, start by
+creating an abstract class with an abstract method (let's call it simply `configure()`) that returns
+the instrumented object - like a client, server, the main class of the instrumented library. Then,
+depending on the chosen test library, go to the [JUnit](#junit) or [Spock](#spock) section.
+
+After writing a test or two, go back to the `library` package and make sure it has
+a `testImplementation` dependency on the `testing` submodule. Then, create a test class that extends
+the abstract test class from `testing`. You should implement the abstract `configure()` method to
+initialize the library using the exposed mechanism to register interceptors/listeners, perhaps a
+method like `registerInterceptor`; or wrap the object with the instrumentation decorator. Make sure
+that proper testing strategy is applied (depends on used test library; but both of them expose a way
+to specify whether you're running a library or javaagent test). If the tests pass, library
+instrumentation is working OK.
+
+### JUnit
+
+The `testing-common` module exposes several JUnit extensions that facilitate writing instrumentation
+tests. In particular, we'll take a look at `LibraryInstrumentationExtension`
+, `AgentInstrumentationExtension` and their parent class `InstrumentationExtension`. The extension
+class implements several useful methods (like `waitAndAssertTraces`, `waitAndAssertMetrics`) that
+you can use in your test cases to verify that the correct telemetry has been produced.
+
+Assuming that this is your abstract test case class:
+
+```java
+public abstract class AbstractYarpcTest {
+
+  protected abstract InstrumentationExtension testing();
+
+  protected abstract Yarpc configure(Yarpc yarpc);
+
+  @Test
+  void testSomething() {
+    // ...
+  }
+}
+```
+
+Aside from the `configure()` method mention in earlier paragraphs, you'll have an
+additional `testing()` method that returns an `InstrumentationExtension` and is supposed to be
+implemented by the extending class.
+
+Now, the library instrumentation class will look like that:
+
+```java
+class LibraryYarpcTest extends AbstractYarpcTest {
+
+  @RegisterExtension
+  InstrumentationExtension testing = LibraryInstrumentationExtension.create();
+
+  @Override
+  protected InstrumentationExtension testing() {
+    return testing;
+  }
+
+  @Override
+  protected Yarpc configure(Yarpc yarpc) {
+    // register interceptor/listener etc
+  }
+}
+```
+
+You can use the JUnit's `@RegisterExtension` annotation to make sure that the instrumentation
```

Comment:
```suggestion
You can use the `@RegisterExtension` annotation of JUnit to make sure that the instrumentation
```

Reply:
Now that I think of it, I can probably remove JUnit here since it's mentioned at the end of the sentence.
```suggestion
You can use the `@RegisterExtension` annotation to make sure that the instrumentation
```